File: src/ds2/LinkedList/SinglyLinkedList/LinkedList.java
package ds2.LinkedList.SinglyLinkedList;

public class LinkedList {
    Node head;


    public LinkedList() {
        this.head = null;
    }

    public LinkedList(Node head) {
        this.head = head;
    }

    public void printList() {
        Node curr = this.head;
        System.out.print("printList: ");
        while(curr != null) {
            System.out.print(curr.data + " ");
            curr = curr.next;
        }
        System.out.println();
    }

    public int size() {
        Node curr = this.head;
        int i = 0;
        while(curr != null) {
            curr = curr.next;
            ++i;
        }
        return i;
    }

    public boolean isEmpty() {
        // return this.size() == 0;
        return this.head == null;
    }

    public int value_at(int index) {
        int count = index;
        Node curr = this.head;
        while(curr != null && count > 0) {
            count--;
            curr = curr.next;
        }
        return curr.data;
    }

    public void pushFront(int value) {
        Node newNode = new Node(value);
        newNode.next = head;
        head = newNode;
    }

    public void pushBack(int value) {
        Node newNode = new Node(value);
        if (head == null)
            head = newNode;
        else {
            Node curr = this.head;
            while(curr.next != null) {
                curr = curr.next;
            }
            curr.next = newNode;
        }
    }

    public int popBack() {
        if (this.isEmpty()) {
            System.out.println("Error: List is empty");
        }
        else {
            Node curr = this.head;
            Node prev = null;
            while(curr.next != null) {
                prev = curr;
                curr = curr.next;
            }
            prev.next = null;
            return curr.data;
        }
        return -1;
    }

    public int frontVal() {
        if (!this.isEmpty()) {
            return this.head.data;
        }
        System.out.println("Error: List is Empty");
        return -1;
    }

    public int back() {
        if (!this.isEmpty()) {
            Node curr = this.head;
            while(curr.next != null) {
                curr = curr.next;
            }
            return curr.data;
        }
        System.out.println("Error: List is Empty");
        return -1;
    }

    public void insertNode(int index, int value) {
        Node newNode = new Node(value);
        int size = this.size();
        if (index > size) {
            System.out.println("Error: index is larger than the size of the list");
        } else if (index == 0) {
            pushFront(value);
        } else if (index == size) {
            pushBack(value);
        } else {

            Node curr = head;
            Node prev = null;
            int i = 0;
            while(curr.next != null && i != index) {
                prev = curr;
                curr = curr.next;
                i++;
            }
            prev.next = newNode;
            newNode.next = curr;
        }
    }

    public void erase(int index) {
        
    }

}

File: DS-And-Algs-Practice-Python/data_structures/LinkedList/LinkedList.py
# I didn't finish but I pretty much know Linked Lists very well

class LinkedList:
    def __init__(self):
        self.head = None

    def size(self):
        curr = self.head
        size = 0
        while curr is not None:
            size += 1
            curr = curr.next
        print("size", size)
        return size

    def empty(self):
        isEmpty = self.head is None
        print("empty", isEmpty)

    def value_at(self, index):
        if index + 1 > self.size():
            print("error: index is too high")
        else:
            curr = self.head
            i = 0
            while curr.next is not None and i is not index:
                i += 1
                curr = curr.next
            print('value_at', curr.val)

    def push_front(self, val):
        newNode = Node(val)
        if self.head is None:
            self.head = newNode
        else:
            newNode.next = self.head
            self.head = newNode

    def pop_front(self):
        if self.head is None:
            print("Linked List is empty")
        else:
            node = self.head
            self.head = self.head.next
            print('pop_front', node.val)
            return node

    def push_back(self, val):
        newNode = Node(val)
        if self.head is None:
            self.head = newNode
        else:
            curr = self.head
            while curr.next is not None:
                curr = curr.next
            curr.next = newNode


    def printList(self):
        curr = self.head
        while curr is not None:
            print(curr.val)
            curr = curr.next

    def pop_back(self):
        if self.head is None:
            print("LinkedList is empty")
        elif self.head.next is None:
            node = self.head
            self.head = None
            print('pop_back', node.val)
            return node
        else:
            curr = self.head
            prev = None
            while curr.next is not None:
                prev = curr
                curr = curr.next
            prev.next = None
            print('pop_back', curr.val)
            return curr

    def front(self):
        if self.head is None:
            print("Linked List is empty")
        else:
            print('front', self.head.val)
        return self.head

    def back(self):
        curr = self.head
        if curr is None:
            print('Linked List is empty')
        else:
            while curr.next is not None:
                curr = curr.next
            print('back', curr.val)
            return curr

class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


ll = LinkedList()
ll.empty()
ll.push_back(1)
ll.push_back(2)
ll.push_back(3)
ll.push_back(4)
ll.push_back(5)
ll.printList()
ll.size()
ll.empty()
ll.value_at(3)
ll.push_front(0)
ll.printList()
ll.pop_front()
ll.printList()
ll.pop_back()
ll.printList()
ll.front()
ll.back()
